### PR TITLE
adding reducer extension types to dynamic-plugin-sdk

### PR DESCRIPTION
**Pull request by glekner**

### CODE REVIEW

**Comment by glekner** [comment]
@vojtechszocs @christianvogt please review

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { Reducer } from 'redux';
+import { CodeRef, Extension } from '@console/plugin-sdk/src/typings/base';
+import { EncodedCodeRef, UpdateExtensionProperties } from '../types';
+
+export namespace ExtensionProperties {
+  export type ReduxReducer = {
+    /** The key to represent reducer-managed substate within the Redux state object. */
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:7]:
```suggestion
    /** The key to represent the reducer-managed substate within the Redux state object. */
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    namespace: string;
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:8]:
```suggestion
    scope: string;
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    /** The reducer handler, resolved to a reducer. */
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:9]:
```suggestion
    /** The reducer function, operating on the reducer-managed substate. */
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    handler: EncodedCodeRef;
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:10]:
```suggestion
    reducer: EncodedCodeRef;
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  };
+  export type ReduxReducerCodeRef = {
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:12]:
```suggestion

  export type ReduxReducerCodeRefs = {
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    /** The reducer handler, resolved to a reducer. */
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:13]:
```suggestion
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    handler: CodeRef<Reducer>;
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:14]:
```suggestion
    reducer: CodeRef<Reducer>;
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  };
+}
+
+// Extension types
+
+export type ReduxReducer = Extension<ExtensionProperties.ReduxReducer> & {
+  type: 'console.reducer';
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:21]:
```suggestion
  type: 'console.redux-reducer';
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+export type ResolvedReduxReducer = UpdateExtensionProperties<
+  ReduxReducer,
+  ExtensionProperties.ReduxReducerCodeRef
+>;
+
+// Type guards
+
+export const isReduxReducer = (e: Extension): e is ResolvedReduxReducer =>
+  e.type === 'console.reducer';
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/reducers.ts:32]:
```suggestion
  e.type === 'console.redux-reducer';
```

File: frontend/packages/console-dynamic-plugin-sdk/src/schema/console-extensions.ts [hunk 10]
```diff
@@ -1,6 +1,7 @@
 import { FeatureFlag, ModelFeatureFlag } from '../extensions/feature-flags';
+import { ReduxReducer } from '../extensions/reducers';
```
Comment (vojtechszocs) [inline frontend/packages/console-dynamic-plugin-sdk/src/schema/console-extensions.ts:2]:
```suggestion
import { ReduxReducer } from '../extensions/redux';
```

**Review by vojtechszocs — CHANGES_REQUESTED** [review]

**Comment by vojtechszocs** [comment]
Looks great, just a couple of comments.

If an extension supports code refs, there should be two extension properties types:
- `<ExtName>` e.g. `ReduxReducer` - contains _all the properties_, with code refs represented as `EncodedCodeRef` (object literals, suitable for JSON usage)
- `<ExtName>CodeRefs` e.g. `ReduxReducerCodeRefs` - contains _only code ref properties_, typed as `CodeRef<T>` where `T` is the actual object being referenced

For such extensions, there should also be two extension types:
- `<ExtName>` e.g. `ReduxReducer`
- `Resolved<ExtName>` e.g. `ResolvedReduxReducer` - using `UpdateExtensionProperties` helper to replace all encoded code refs with their actual runtime types

When documenting such extension's properties, just document the base type. The `<ExtName>CodeRefs` type is a subset of the base type, so no need to document that.

Also, type guards (`is<ExtName>` e.g. `isReduxReducer`) for such extensions should work with the _resolved_ type, since we assume that code refs will be resolved _before_ the extension instance is interpreted (preferably via `useResolvedExtensions` hook).

**Comment by spadgett** [comment]
/override ci/prow/ceph-storage-plugin

The ceph-storage-plugin job should only have run on PRs with changes to `frontend/packages/ceph-storage-plugin`. It failed due to an unrelated CI config change. Overriding status on this PR.

**Comment by glekner** [comment]
can you lgtm @vojtechszocs

**Pull request merged**